overlord/servicestate/servicestate.go

```diff
@@ -42,14 +44,185 @@ import (
 	"github.com/snapcore/snapd/wrappers"
 )
 
+type UserSelection int
+
+const (
+	UserSelectionList UserSelection = iota
+	UserSelectionSelf
+	UserSelectionAll
+)
+
+// UserSelector is a support structure for correctly translating a way of
+// representing both a list of user-names, and specific keywords like "self"
+// and "all" for JSON marshalling.
+//
+// When "Selector == UserSelectionList" then Names is used as the data source and
+// the data is treated like a list of strings.
+// When "Selector == UserSelectionSelf|UserSelectionAll", then the data source will
+// be a single string that represent this in the form of "self|all".
+type UserSelector struct {
+	Names    []string
+	Selector UserSelection
+}
+
+// UserList returns a decoded list of users which takes any keyword into account.
+// Takes the current user to be able to handle special keywords like 'user'.
+func (us *UserSelector) UserList(currentUser *user.User) ([]string, error) {
+	switch us.Selector {
+	case UserSelectionList:
+		return us.Names, nil
+	case UserSelectionSelf:
+		if currentUser == nil {
+			return nil, fmt.Errorf("internal error: for \"self\" the current user must be provided")
+		}
+		if currentUser.Uid == "0" {
+			return nil, fmt.Errorf("cannot use \"self\" for root user")
+		}
+		return []string{currentUser.Username}, nil
+	case UserSelectionAll:
+		// Empty list indicates all.
+		return nil, nil
+	}
+	return nil, fmt.Errorf("internal error: unsupported selector %d specified", us.Selector)
+}
+
+func (us UserSelector) MarshalJSON() ([]byte, error) {
+	switch us.Selector {
+	case UserSelectionList:
+		return json.Marshal(us.Names)
+	case UserSelectionSelf:
+		return json.Marshal("self")
+	case UserSelectionAll:
+		return json.Marshal("all")
+	default:
+		return nil, fmt.Errorf("internal error: unsupported selector %d specified", us.Selector)
+	}
+}
+
+func (us *UserSelector) UnmarshalJSON(b []byte) error {
+	// Try treating it as a list of usernames first
+	var users []string
+	if err := json.Unmarshal(b, &users); err == nil {
+		us.Names = users
+		us.Selector = UserSelectionList
+		return nil
+	}
+
+	// Fallback to string, which would indicate a keyword
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	switch s {
+	case "self":
+		us.Selector = UserSelectionSelf
+	case "all":
+		us.Selector = UserSelectionAll
+	}
+	return nil
+}
+
 type Instruction struct {
-	Action string   `json:"action"`
-	Names  []string `json:"names"`
+	Action string       `json:"action"`
+	Names  []string     `json:"names"`
+	Scope  []string     `json:"scope"`
```

Comment:
Looking at ServiceScope() below, looks like this could really be:

```
type ScopeSelector wrappers.ServiceScope

func (s *ScopeSelector) UnmarshalJSON(d []byte) error {
     // figure out the scope

}
```
conveniently wrappers.ServiceScope is a string, so if none is provided in the request we'd get `""` so that the check in EnsureDefaultScopeForUser() can still be done, just this time on `Scope == ""`

Reply:
So this made a lot of sense and definitely did an improvement, I really wanted to do this but did not want to add Unmarshal code to the wrappers.ServiceScope, so thank you for learning me that I can retype it and add unmarshal code for that instead! This worked great.